lib/protocol/protocol.go

```diff
@@ -823,26 +753,63 @@ func (c *rawConnection) writeUncompressedMessage(msg message) error {
 
 	totSize := 2 + hdrSize + 4 + size
 	buf := BufferPool.Get(totSize)
+	defer BufferPool.Put(buf)
 
 	// Header length
 	binary.BigEndian.PutUint16(buf, uint16(hdrSize))
 	// Header
 	if _, err := hdr.MarshalTo(buf[2:]); err != nil {
-		BufferPool.Put(buf)
 		return errors.Wrap(err, "marshalling header")
 	}
 	// Message length
 	binary.BigEndian.PutUint32(buf[2+hdrSize:], uint32(size))
 	// Message
 	if _, err := msg.MarshalTo(buf[2+hdrSize+4:]); err != nil {
-		BufferPool.Put(buf)
 		return errors.Wrap(err, "marshalling message")
 	}
```

Comment:
This part is has no effect/is unneccessary if the message is sent compressed.

Reply:
It does (except for header marshaling), because this determines what needs to be compressed. The original code also marshaled twice in the compressed case.